workshops/views.py

```diff
@@ -145,51 +148,201 @@ def person_bulk_add(request):
+    """
+    Read data from CSV and turn it into JSON-serializable list of dictionaries.
+    "Serializability" is required because we put this data into session.  See
+    https://docs.djangoproject.com/en/1.7/topics/http/sessions/ for details.
+    """
     persons_tasks = []
     reader = csv.DictReader(uploaded_file)
     for row in reader:
-        person_fields = dict((col, row[col].strip()) for col in PERSON_UPLOAD_FIELDS)
-        person = Person(**person_fields)
-        entry = {'person': person, 'task' : None}
-        if row.get('event', None) and row.get('role', None):
+        person_fields = dict((col, row[col].strip())
+                             for col in PERSON_UPLOAD_FIELDS)
+        entry = {'person': person_fields, 'event': row.get('event', None),
+                 'role': row.get('role', None), 'errors': None}
+        persons_tasks.append(entry)
+    return persons_tasks
+
+
+def _verify_upload_person_task(data):
+    """
+    Verify that uploaded data is correct.  Show errors by populating ``errors``
+    dictionary item.
+
+    This function changes ``data`` in place; ``data`` is a list, so it's
+    passed by a reference.  This means any changes to ``data`` we make in
+    here don't need to be returned via ``return`` statement.
+    """
+
+    errors_occur = False
+
+    for item in data:
+        event, role = item.get('event', None), item.get('role', None)
+        email = item['person'].get('email', None)
+
+        errors = []
+        if event:
             try:
-                event = Event.objects.get(slug=row['event'])
-                role = Role.objects.get(name=row['role'])
-                entry['task'] = Task(person=person, event=event, role=role)
-                import sys
+                Event.objects.get(slug=event)
             except Event.DoesNotExist:
-                messages.add_message(request, messages.ERROR, \
-                                     'Event with slug {} does not exist.'.format(row['event']))
+                errors.append('Event with slug {} does not exist.'
+                              .format(event))
+        if role:
+            try:
+                Role.objects.get(name=role)
             except Role.DoesNotExist:
-                messages.add_message(request, messages.ERROR, \
-                                     'Role with name {} does not exist.'.format(row['role'])) 
+                errors.append('Role with name {} does not exist.'.format(role))
             except Role.MultipleObjectsReturned:
-                messages.add_message(request, messages.ERROR, \
-                                     'More than one role named {} exists.'.format(row['role'])) 
-        persons_tasks.append(entry)
-    return persons_tasks
+                errors.append('More than one role named {} exists.'
+                              .format(role))
+
+        if email:
+            try:
+                Person.objects.get(email=email)
```

Comment:
Email addresses aren't case-sensitive, so to ensure you don't have differently-cased but otherwise the same address, you should do `Person.objects.get(email__iexact=email)`.